rasterio/__init__.py

```diff
@@ -219,7 +222,7 @@ def fp_reader(fp):
 
         return fp_reader(fp)
 
-    elif mode == 'w' and hasattr(fp, 'write'):
+    elif mode.startswith("w") and hasattr(fp, 'write'):
```

Comment:
I might be wrong but what happens if someone try 
```
with rasterio.open('a.tif', 'r') as f:
    meta = f.meta.copy()
    with rasterio.open('another.tif', 'wa', **meta) as ff:
        ff.write(f.read())
```
to me it seems the `mode.startswith("w")` could be too permissive 

though @sgillies ?